enterprise/kernel/src/main/java/org/neo4j/kernel/impl/pagecache/PageCacheWarmer.java

```diff
@@ -113,16 +104,22 @@ public synchronized Resource addFilesTo( Collection<StoreFileMetadata> coll ) th
         return Resource.EMPTY;
     }
 
+    @Override
+    public synchronized void start()
+    {
+        stopped = false;
+        executor = buildExecutorService( scheduler );
+        pageLoaderFactory = new PageLoaderFactory( executor, pageCache );
+    }
+
+    @Override
     public void stop()
     {
         stopped = true;
         synchronized ( this )
         {
-            // This synchronised block means we'll wait for any reheat() or profile() to notice our raised stopped flag,
-            // before we return to the caller. This helps avoid races, e.g. if the page cache is closed while this page
-            // cache warmer is still holding on to some mapped pages.
+            executor.shutdown();
```

Comment:
Still looks strange since your synchronizing on a thread-safe object. Someone coming here might believe that you can remove that block, completely missing the intended barrier this synchronize provide.

I would add an additional variable that indicates that the profile/reheat is running and wait for that to flip to make this intention clear.
```
        // wait for shutdown
        synchronized ( this ) {
            while ( working ) {
                wait();
            }
        }
```